### PR TITLE
[WIP] Intrinsics now platform dependent. LibC::PtrdiffT type introduced.

### LINKED ISSUE #4589: Pointer's copy_from|copy_to|move_from|move_to are inherently broken.
Following lines certainly lead to wrong behavior on 64bit platforms with large arrays cause of `#to_u32`.

https://github.com/crystal-lang/crystal/blob/46e9d81b965d9e00dea4aeb0002999c756545b6a/src/pointer.cr#L245
https://github.com/crystal-lang/crystal/blob/46e9d81b965d9e00dea4aeb0002999c756545b6a/src/pointer.cr#L258
https://github.com/crystal-lang/crystal/blob/46e9d81b965d9e00dea4aeb0002999c756545b6a/src/pointer.cr#L499

BTW, why Crystal has neither `size_t` nor `intptr_t` equivalent? Some places uses `UInt64`, some uses `UInt32` (like this one). It is just... infantile.

For example, Go defines `int` to be `ssize_t` and `uint` to be `size_t`, and has `intptr` and `uintptr` (though, looks like on all supported platforms they have same size to `int`).
Having platform dependent `int` has profit in low-level programming, and dis-convenient for application level programming, so it certainly should not be default type.
But still any place that deals with low-level memory handling should operate on platform dependent types/aliases instead of raw `UInt32` and/or `UInt64`. This bogus lines are just tip of iceberg.

### PR DESCRIPTION
**Pull request by akzhan**
Intrinsics now declared per architecture.

LibC::PtrdiffT is the signed integer type of the result of subtracting two pointers.

LibC::PtrdiffT is used for pointer arithmetic and array indexing, if negative values are possible.

Refs #823, Closes #4589.

/cc @ysbaddaden

### CODE REVIEW

**Comment by akzhan** [comment]
Got failure on 32bit build. Will check.

File: src/lib_c/aarch64-linux-gnu/c/intrinsics.cr [hunk 1]
```diff
@@ -0,0 +1,20 @@
+lib Intrinsics
```
Comment (ysbaddaden) [inline src/lib_c/aarch64-linux-gnu/c/intrinsics.cr]: Intrinsics aren't LibC bindings (i.e. calls to external symbols) but LLVM intrinsics (inlined calls); they won't be generated automatically; they aren't system dependent (but arch dependent).

I.e.: they have nothing to do under `LibC`, please put them back as `src/intrinsics.cr` or maybe `src/crystal/intrinsics.cr`, along with the flag macro to select either the u32 or u64 calls.

File: src/pointer.cr [hunk 2]
```diff
@@ -242,7 +242,7 @@ struct Pointer(T)
     raise ArgumentError.new("Negative count") if count < 0
 
     if self.class == source.class
-      Intrinsics.memcpy(self.as(Void*), source.as(Void*), (count * sizeof(T)).to_u32, 0_u32, false)
+      Intrinsics.memcpy(self.as(Void*), source.as(Void*), Intrinsics::SizeT.new(count) * sizeof(T), 0_u32, false)
```
Comment (ysbaddaden) [inline src/pointer.cr]: Why introduce `Intrinsics::SizeT` but use `LibC::SizeT` below? This is confusing *and* a duplication.

File: src/pointer.cr [hunk 3]
```diff
@@ -395,7 +395,7 @@ struct Pointer(T)
   # ptr.address # => 0
   # ```
   def self.null
-    new 0_u64
+    new LibC::SizeT.new(0)
```
Comment (ysbaddaden) [inline src/pointer.cr]: 1. `size_t` is a length type, you meant `intptr_t` (pointer size);
2. you must hack the compiler for this to compile on 32bit platforms (`Pointer.new(UInt64)` is a Crystal intrinsic).

File: src/pointer.cr [hunk 4]
```diff
@@ -405,7 +405,7 @@ struct Pointer(T)
   # ptr.address # => 5678
   # ```
   def self.new(address : Int)
-    new address.to_u64
+    new LibC::SizeT.new(address)
```
Comment (ysbaddaden) [inline src/pointer.cr]: Same here.

File: src/pointer.cr [hunk 5]
```diff
@@ -78,7 +78,7 @@ struct Pointer(T)
   # ptr2.address # => 1238
   # ```
   def +(other : Int)
-    self + other.to_i64
+    self + LibC::PtrdiffT.new(other)
```
Comment (ysbaddaden) [inline src/pointer.cr]: You must hack the compiler for this to compile on 32bit platforms (`Pointer#+(Int64)` is a Crystal intrinsic). See below.

File: src/lib_c/aarch64-linux-gnu/c/intrinsics.cr (hunk 1 above)
Comment (akzhan) [inline src/lib_c/aarch64-linux-gnu/c/intrinsics.cr]: I like idea of separated files per arch. Will try to rework it accordingly.

File: src/pointer.cr (hunk 5 above)
Comment (akzhan) [inline src/pointer.cr]: Thanks for the explanation.

File: src/pointer.cr (hunk 3 above)
Comment (akzhan) [inline src/pointer.cr]: Thanks, it was a big mistake.

**Comment by akzhan** [comment]
I should close this because another approach has been landed - #4960.

**Pull request closed without merging**